### PR TITLE
L2-483: Lazy load voted proposals in Neuron Voting History

### PR DESCRIPTION
**Pull request by lmuntanerDfinity**
# Motivation

Do not fetch from the backend the proposal for each ballot that the neuron has voted, until the user needs to see it.

# Changes

* Use the `InfiniteScroll` in the `Ballots` components to load ballots per page.
* Add two props in `InfiinteScroll` for new use case: `pageLimit` and `containerElement`.

# Tests

* New test cases in InfiniteScroll

### CODE REVIEW

**Comment by lmuntanerDfinity** [comment]
UX Videos:

Scrolling slowly to see the sequential calls to get proposals.

https://user-images.githubusercontent.com/98735375/166415843-1201b233-8bd0-4de7-8407-cefcadda9a33.mov

Scrolling fast and faking a fetch.

https://user-images.githubusercontent.com/98735375/166415869-c31730c4-0b3a-4a29-bdcd-af05205fb747.mov

**Comment by lmuntanerDfinity** [comment]
@mstrasinskis @peterpeterparker please review

File: frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte [hunk 1]
```diff
@@ -101,6 +102,20 @@
   onDestroy(() => observer.disconnect());
 </script>
 
-<div bind:this={container}>
-  <slot />
-</div>
+{#if containerElement === "ul"}
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte:105]: do you absolutely need a `ul`? if yes, why?

If not absolutely needed, I feel like it would be cleaner to not add such logic to the InfiniteScroll component

File: frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte [hunk 2]
```diff
@@ -3,11 +3,34 @@
...
+  // We want to avoid making too many calls, since a neuron can vote in many proposals.
+  let ballotsToShow: Required<BallotInfo>[] = [];
+  let ballotsIndex: number = PAGE_LIMIT;
   $: ballots = neuron === undefined ? [] : ballotsWithDefinedProposal(neuron);
+  $: ballotsToShow = ballots.slice(0, ballotsIndex);
+
+  // We fake fetching the next `PAGE_LIMIT` ballots.
+  const nextPage = debounce(() => {
+    ballotsIndex += PAGE_LIMIT;
+    fakeLoading = false;
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte:24]: should the `fakeLoading` be set back to `false` only when the new index is applied to `ballotsToShow`?

File: frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte (hunk 1 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte:105]: I agree, I guess it's a question of html semantics.

In the list of votes, it made sense to use a `ul` and `li` for each item in the vote history. Since that is a list.

If we are ok with not using a `ul` and `li` for this list of votings, like we do for `Proposals`, then I can remove this logic.

What do you prefer?

File: frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte (hunk 2 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte:24]: Why? I find it easier to read here, but if there is a good reason to put it in the side effect, I can put it there as well.

File: frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte (hunk 1 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte:105]: Sure why not, not against. As long as there is not such `containerElement` logic in the InfiniteScoll component I am happy 😃

File: frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte (hunk 2 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte:24]: the infinite scroll might trigger an interesection again while the ballots are not rendered, so it might already fetch the next ballots even though the user did not scroll to read the last ballots that have been fetched. edge case. not that bad though.

File: frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte (hunk 1 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte:105]: Shall I change the Proposals and use always a `ul`? Do I do that in this PR?

Not sure I understand what you want me to do.

File: frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte (hunk 2 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte:24]: yes, specially not for this case.

File: frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte (hunk 1 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte:105]: ok

Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte]: done

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker ready for next review

File: frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte (hunk 2 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/components/neuron-detail/Ballots/Ballots.svelte:24]: right, let's see if detect edge cases during test phase

File: frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte [hunk 3]
```diff
@@ -19,7 +21,7 @@
     threshold: 0,
   };
 
-  let container: HTMLDivElement;
+  let container: HTMLDivElement | HTMLUListElement;
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/InfiniteScroll.svelte:24]: it's always a `HTMLUListElement` now

File: frontend/svelte/src/routes/Proposals.svelte [hunk 4]
```diff
@@ -156,7 +156,9 @@
 
       <InfiniteScroll on:nnsIntersect={findNextProposals}>
         {#each $proposalsStore.proposals as proposalInfo (proposalInfo.id)}
-          <ProposalCard {hidden} {proposalInfo} />
+          <li>
```
Comment (peterpeterparker) [inline frontend/svelte/src/routes/Proposals.svelte:159]: the `li` should be injected / used within `ProposalCard` and replaces the `<div class:hidden>` otherwise we will have issue with one of the filter of the "Voting" tab (the one that happens only on the fronend side)

**Review by peterpeterparker — CHANGES_REQUESTED** [review]
The comment `li` is important in my opinion.

File: frontend/svelte/src/routes/Proposals.svelte (hunk 4 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/routes/Proposals.svelte:159]: I thought that the ProposalCard was used somewhere else, but it's only used here, so I can change it to `li`.

Comment (lmuntanerDfinity) [reply frontend/svelte/src/routes/Proposals.svelte]: done

**Comment by lmuntanerDfinity** [comment]
The change is important indeed, @peterpeterparker ready for next review

**Review by peterpeterparker — APPROVED** [review]
LGTM

As discussed, sync the merge with release, thanks a lot!
